src/occupation.jl

```diff
@@ -1,61 +1,99 @@
-# Functions for finding the Fermi level and occupation numbers for bands
-
 import Roots
 
+# Functions for finding the Fermi level and occupation numbers for bands
+# The goal is to find εF so that
+# f_i = filled_occ * f((εi-εF)/θ) with θ = temperature
+# sum_i f_i = n_electrons
+# If temperature is zero, (εi-εF)/θ = ±∞.
+# The occupation function is required to give 1 and 0 respectively in these cases.
+#
+# For finite temperature we right now just use bisection; note that with MP smearing
+# there might be multiple possible Fermi levels. This could be sped up
+# with more advanced methods (e.g. false position), but more care has to
+# be taken with convergence criteria and the like
 
-"""
-Find the occupation and Fermi level.
-"""
-function compute_occupation(basis::PlaneWaveBasis, eigenvalues;
-                            temperature=basis.model.temperature,
-                            smearing=basis.model.smearing)
-    if !isnothing(basis.model.εF)  # fixed Fermi level
-        εF = basis.model.εF
-    else  # fixed n_electrons
-        εF = compute_fermi_level(basis, eigenvalues; temperature, smearing)
-    end
-    occupation = compute_occupation(basis, eigenvalues, εF; temperature, smearing)
+abstract type FermiAlgorithm end
 
-    (; occupation, εF)
-end
+"""Default selection of a Fermi level determination algorithm"""
+default_fermialg(::Type{<: Smearing.SmearingFunction}) = FermiTwoStage()
+default_fermialg(::Type{Smearing.Gaussian})   = FermiBisection()  # Monotonic smearing
+default_fermialg(::Type{Smearing.FermiDirac}) = FermiBisection()  # Monotonic smearing
+default_fermialg(::T) where {T <: Smearing.SmearingFunction} = default_fermialg(T)
+default_fermialg(model::Model) = default_fermialg(model.smearing)
+default_fermialg(basis::AbstractBasis) = default_fermialg(basis.model)
```

Comment:
that's a lot of methods... Might as well remove the last one (basis). Why don't you also just dispatch on instances instead of types?